### PR TITLE
fix(lodepng): fix LodePNG 8 bit palette image to RGBA data conversion bug

### PR DESCRIPTION
**Pull request by PGNetHun**
### Description of the feature or fix

A new bug was introduced in PR https://github.com/lvgl/lvgl/pull/4968:
decoding 8 bit palette PNG file resulted memory violation exception, due to converting from palette colors to RGBA was not changed (and not tested) to use the new approach (using `lv_draw_buf_t` instead of raw image data).
The bug is fixed, and LodePNG library test is also extended to test decoding of a 8 bit palette PNG file.

### Checkpoints
- [x] Run `code-format.py` from the scripts folder. [astyle](http://astyle.sourceforge.net/install.html) needs to be installed.
- [ ] Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed
- [ ] Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant. 
- [x] Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- [ ] If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/release/v8.3/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/release/v8.3/Kconfig).
 
Be sure the following conventions are followed:
- [x] Follow the [Styling guide](https://github.com/lvgl/lvgl/blob/master/docs/CODING_STYLE.md)
- [ ] Prefer `enum`s instead of macros. If inevitable to use `define`s export them with `LV_EXPORT_CONST_INT(defined_value)` right after the `define`.
- [ ] In function arguments prefer `type name[]` declaration for array parameters instead of `type * name`
- [ ] Use typed pointers instead of `void *` pointers
- [ ] Do not `malloc` into a static or global variables. Instead declare the variable in `lv_global_t` structure in [`lv_global.h`](https://github.com/lvgl/lvgl/blob/master/src/core/lv_global.h) and mark the variable with `(LV_GLOBAL_DEFAULT()->variable)` when it's used. See a detailed description [here](https://docs.lvgl.io/master/get-started/bindings/micropython.html#memory-management).
- [ ] Widget constructor must follow the `lv_<widget_name>_create(lv_obj_t * parent)` pattern.
- [ ] Widget members function must start with `lv_<module_name>` and should receive `lv_obj_t *` as first argument which is a pointer to widget object itself.  
- [ ] `struct`s should be used via an API and not modified directly via their elements.
- [ ] `struct` APIs should follow the widgets' conventions. That is to receive a pointer to the `struct` as the first argument, and the prefix of the `struct` name should be used as the prefix of the function name too (e.g.  `lv_disp_set_default(lv_disp_t * disp)`)
- [ ] Functions and `struct`s which are not part of the public API must begin with underscore in order to mark them as "private".
- [ ] Arguments must be named in H files too.
- [ ] To register and use callbacks one of the following needs to be followed (see a detailed description [here](https://docs.lvgl.io/master/get-started/bindings/micropython.html#callbacks)):
  - For both the registration function and the callback pass a pointer to a `struct` as the first argument. The `struct` must contain `void * user_data` field.
  - The last argument of the registration function must be `void * user_data` and the same `user_data` needs to be passed as the last argument of the callback.
  - Callback types not following these conventions should end with `xcb_t`.

### CODE REVIEW

**Review by XuNeo — APPROVED** [review]
Thank you!

File: src/libs/lodepng/lodepng.c [hunk 1]
```diff
@@ -5349,14 +5348,23 @@ unsigned lodepng_decode(unsigned char ** out, unsigned * w, unsigned * h,
             return 56; /*unsupported color mode conversion*/
         }
 
-        outsize = lodepng_get_raw_size(*w, *h, &state->info_raw);
-        *out = (unsigned char *)lodepng_malloc(outsize);
-        if(!(*out)) {
+        lv_draw_buf_t * new_buf = lv_draw_buf_create(*w, *h, LV_COLOR_FORMAT_ARGB8888, 4 * *w);
+        if(new_buf == NULL) {
             state->error = 83; /*alloc fail*/
+            return state->error;
```
Comment (terry0012) [inline src/libs/lodepng/lodepng.c]: should we free out buffer as it maybe not NULL in decodeGeneric and set it to NULL ?

File: src/libs/lodepng/lodepng.c [hunk 2]
```diff
@@ -5349,14 +5348,23 @@ unsigned lodepng_decode(unsigned char ** out, unsigned * w, unsigned * h,
...
         }
-        else state->error = lodepng_convert(*out, data, &state->info_raw,
+
+        state->error = lodepng_convert(new_buf->data, old_buf->data, &state->info_raw,
                                                 &state->info_png.color, *w, *h);
-        lodepng_free(data);
+
+        if (state->error) {
+            lv_draw_buf_destroy(new_buf);
```
Comment (terry0012) [inline src/libs/lodepng/lodepng.c]: In this case, new_buf is freed, and old_buf is freed later, but out value is not NULL.
should we set it to NULL like *out = NULL ?

Comment (PGNetHun) [reply src/libs/lodepng/lodepng.c]: Yes, you are correct!

File: src/libs/lodepng/lodepng.c (hunk 1 above)
Comment (PGNetHun) [reply src/libs/lodepng/lodepng.c]: Fixed memory leak in case decoding fails.

File: src/libs/lodepng/lodepng.c (hunk 2 above)
Comment (PGNetHun) [reply src/libs/lodepng/lodepng.c]: Fixed memory leak in case decoding fails.

File: src/libs/lodepng/lodepng.c [hunk 3]
```diff
@@ -5349,14 +5348,25 @@ unsigned lodepng_decode(unsigned char ** out, unsigned * w, unsigned * h,
...
+            *out = NULL;
+        }
+        else {
+            *out = (unsigned char*)new_buf;
+        }
+        lv_draw_buf_destroy(old_buf);
```
Comment (terry0012) [inline src/libs/lodepng/lodepng.c:5366]: why not use the orignal logic， it's more clear and will reduce the risk of memory leak.
``` C
 lv_draw_buf_t * new_buf = lv_draw_buf_create(*w, *h, LV_COLOR_FORMAT_ARGB8888, 4 * *w);

 if(new_buf != NULL) {
    state->error = lodepng_convert(new_buf->data, old_buf->data, &state->info_raw,
                                                &state->info_png.color, *w, *h);
    if (state->error) {
        lv_draw_buf_destroy(new_buf);
        new_buf = NULL;
    } 
 }

 *out = (unsigned char*)new_buf;
 lv_draw_buf_destroy(old_buf);

Comment (PGNetHun) [reply src/libs/lodepng/lodepng.c:5366]: Thanks for the productive feedback!
Code is updated.

**Review by terry0012 — APPROVED** [review]

File: tests/src/test_cases/libs/test_lodepng.c [hunk 4]
```diff
@@ -30,13 +31,23 @@ static void create_images(void)
...
-    lv_obj_align(img, LV_ALIGN_CENTER, 100, -20);
+    lv_obj_align(img, LV_ALIGN_CENTER, 100, -100);
 
     label = lv_label_create(lv_screen_active());
-    lv_label_set_text(label, "File");
-    lv_obj_align(label, LV_ALIGN_CENTER, 100, 20);
+    lv_label_set_text(label, "File (32 bit)");
+    lv_obj_align(label, LV_ALIGN_CENTER, 100, -60);
+
+    // 8 bit palette PNG file
```
Comment (kisvegabor) [inline tests/src/test_cases/libs/test_lodepng.c]:
```suggestion
    /* 8 bit palette PNG file*/
```

File: tests/src/test_cases/libs/test_lodepng.c [hunk 5]
```diff
@@ -30,13 +31,23 @@ static void create_images(void)
     lv_label_set_text(label, "Array");
     lv_obj_align(label, LV_ALIGN_CENTER, -100, 20);
 
+    // 32 bit PNG file
```
Comment (kisvegabor) [inline tests/src/test_cases/libs/test_lodepng.c]:
```suggestion
    /* 32 bit PNG file */
```

File: tests/src/test_cases/libs/test_lodepng.c [hunk 6]
```diff
@@ -21,6 +21,7 @@ static void create_images(void)
     lv_obj_t * img;
     lv_obj_t * label;
 
+    // PNG array
```
Comment (kisvegabor) [inline tests/src/test_cases/libs/test_lodepng.c]:
```suggestion
    /* PNG array */
```

**Comment by kisvegabor** [comment]
Thank you! Just a little bit of nit-picking about comments style.

File: tests/src/test_cases/libs/test_lodepng.c (hunk 4 above)
Comment (PGNetHun) [reply tests/src/test_cases/libs/test_lodepng.c]: Fixed.

File: tests/src/test_cases/libs/test_lodepng.c (hunk 5 above)
Comment (PGNetHun) [reply tests/src/test_cases/libs/test_lodepng.c]: Fixed.

File: tests/src/test_cases/libs/test_lodepng.c (hunk 6 above)
Comment (PGNetHun) [reply tests/src/test_cases/libs/test_lodepng.c]: Fixed.

**Review by kisvegabor — APPROVED** [review]

**Review by FASTSHIFT — APPROVED** [review]

**Pull request merged**